File: go/hybrid/ecies_aead_hkdf_dem_helper_test.go
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package hybrid

import (
	"bytes"
	"testing"

	"github.com/google/tink/go/aead"
	"github.com/google/tink/go/daead"
	"github.com/google/tink/go/mac"
	"github.com/google/tink/go/signature"
	"github.com/google/tink/go/subtle/random"
	"github.com/google/tink/go/tink"
	tinkpb "github.com/google/tink/go/proto/tink_go_proto"
)

type eciesAEADHKDFDEMHelperTestCase struct {
	name     string
	template *tinkpb.KeyTemplate
	keySize  uint32
}

var (
	eciesAEADHKDFDEMHelperSupportedAEADs = []eciesAEADHKDFDEMHelperTestCase{
		{
			name:     "AESCTRHMACSHA256",
			template: aead.AES256CTRHMACSHA256KeyTemplate(),
			keySize:  64,
		},
		{
			name:     "AES128CTRHMACSHA256",
			template: aead.AES128CTRHMACSHA256KeyTemplate(),
			keySize:  48,
		},
		{
			name:     "AES256GCM",
			template: aead.AES256GCMKeyTemplate(),
			keySize:  32,
		},
		{
			name:     "AES128GCM",
			template: aead.AES128GCMKeyTemplate(),
			keySize:  16,
		},
	}

	eciesAEADHKDFDEMHelperSupportedDAEADs = []eciesAEADHKDFDEMHelperTestCase{
		{
			name:     "AESSIV",
			template: daead.AESSIVKeyTemplate(),
			keySize:  64,
		},
	}
)

func TestECIESAEADHKDFDEMHelper_AEADKeyTemplates(t *testing.T) {
	plaintext := random.GetRandomBytes(20)
	associatedData := random.GetRandomBytes(20)

	for _, tc := range eciesAEADHKDFDEMHelperSupportedAEADs {
		t.Run(tc.name, func(t *testing.T) {
			dem, err := newRegisterECIESAEADHKDFDemHelper(tc.template)
			if err != nil {
				t.Fatalf("newRegisterECIESAEADHKDFDEMHelper(tc.template) err = %s, want nil", err)
			}

			sk := random.GetRandomBytes(dem.GetSymmetricKeySize())
			primitive, err := dem.GetAEADOrDAEAD(sk)
			if err != nil {
				t.Fatalf("dem.GetAEADorDAEAD(sk) err = %v, want nil", err)
			}
			a, ok := primitive.(tink.AEAD)
			if !ok {
				t.Fatalf("primitive is not of type tink.AEAD")
			}

			var ciphertext []byte
			ciphertext, err = a.Encrypt(plaintext, associatedData)
			if err != nil {
				t.Fatalf("a.Encrypt() err = %v, want nil", err)
			}

			var decrypted []byte
			decrypted, err = a.Decrypt(ciphertext, associatedData)
			if err != nil {
				t.Fatalf("a.Decrypt() err = %v, want nil", err)
			}
			if !bytes.Equal(decrypted, plaintext) {
				t.Errorf("a.Decrypt() = %x, want: %x", decrypted, plaintext)
			}
		})
	}
}

func TestECIESAEADHKDFDEMHelper_DAEADKeyTemplates(t *testing.T) {
	plaintext := random.GetRandomBytes(20)
	associatedData := random.GetRandomBytes(20)

	for _, tc := range eciesAEADHKDFDEMHelperSupportedDAEADs {
		t.Run(tc.name, func(t *testing.T) {
			dem, err := newRegisterECIESAEADHKDFDemHelper(tc.template)
			if err != nil {
				t.Fatalf("newRegisterECIESAEADHKDFDEMHelper(tc.template) err = %s, want nil", err)
			}

			sk := random.GetRandomBytes(dem.GetSymmetricKeySize())
			primitive, err := dem.GetAEADOrDAEAD(sk)
			if err != nil {
				t.Fatalf("dem.GetAEADorDAEAD(sk) err = %v, want nil", err)
			}
			d, ok := primitive.(tink.DeterministicAEAD)
			if !ok {
				t.Fatalf("primitive is not of type tink.DeterministicAEAD")
			}

			var ciphertext []byte
			ciphertext, err = d.EncryptDeterministically(plaintext, associatedData)
			if err != nil {
				t.Fatalf("d.Encrypt() err = %v, want nil", err)
			}

			var decrypted []byte
			decrypted, err = d.DecryptDeterministically(ciphertext, associatedData)
			if err != nil {
				t.Fatalf("d.Decrypt() err = %v, want nil", err)
			}
			if !bytes.Equal(decrypted, plaintext) {
				t.Errorf("d.Decrypt() = %x, want: %x", decrypted, plaintext)
			}
		})
	}
}

func TestECIESAEADHKDFDEMHelper_KeySizes(t *testing.T) {
	var testCases []eciesAEADHKDFDEMHelperTestCase
	testCases = append(testCases, eciesAEADHKDFDEMHelperSupportedAEADs...)
	testCases = append(testCases, eciesAEADHKDFDEMHelperSupportedDAEADs...)

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			dem, err := newRegisterECIESAEADHKDFDemHelper(tc.template)
			if err != nil {
				t.Fatalf("newRegisterECIESAEADHKDFDemHelper(tc.template): %s", err)
			}
			if dem.GetSymmetricKeySize() != tc.keySize {
				t.Errorf("dem.GetSymmetricKeySize() = %d, want: %d", dem.GetSymmetricKeySize(), tc.keySize)
			}

			shortKey := make([]byte, tc.keySize-1)
			if _, err = dem.GetAEADOrDAEAD(shortKey); err == nil {
				t.Errorf("dem.GetAEADOrDAEAD(shortKey) err = nil, want non-nil")
			}

			longKey := make([]byte, tc.keySize+1)
			if _, err = dem.GetAEADOrDAEAD(longKey); err == nil {
				t.Errorf("dem.GetAEADOrDAEAD(longKey) err = nil, want non-nil")
			}
		})
	}
}

func TestECIESAEADHKDFDEMHelper_UnsupportedKeyTemplates(t *testing.T) {
	testCases := []struct {
		name     string
		template *tinkpb.KeyTemplate
	}{
		{
			name:     "signature",
			template: signature.ECDSAP256KeyTemplate(),
		},
		{
			name:     "mac",
			template: mac.HMACSHA256Tag256KeyTemplate(),
		},
		{
			name:     "invalid_type_and_value",
			template: &tinkpb.KeyTemplate{TypeUrl: "some url", Value: []byte{0}},
		},
		{
			name:     "aesctrhmac_empty_value",
			template: &tinkpb.KeyTemplate{TypeUrl: aesCTRHMACAEADTypeURL},
		},
		{
			name:     "aesgcm_empty_value",
			template: &tinkpb.KeyTemplate{TypeUrl: aesGCMTypeURL},
		},
		{
			name:     "aessiv_empty_value",
			template: &tinkpb.KeyTemplate{TypeUrl: aesSIVTypeURL},
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			if _, err := newRegisterECIESAEADHKDFDemHelper(tc.template); err == nil {
				t.Errorf("newRegisterECIESAEADHKDFDemHelper() err = nil, want non-nil")
			}
		})
	}
}

File: cc/hybrid/ecies_aead_hkdf_dem_helper_test.cc
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
///////////////////////////////////////////////////////////////////////////////

#include "tink/hybrid/ecies_aead_hkdf_dem_helper.h"

#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "absl/status/status.h"
#include "tink/aead/aes_gcm_key_manager.h"
#include "tink/daead/aes_siv_key_manager.h"
#include "tink/registry.h"
#include "tink/util/secret_data.h"
#include "tink/util/test_matchers.h"
#include "tink/util/test_util.h"

namespace crypto {
namespace tink {
namespace {

using ::crypto::tink::subtle::AeadOrDaead;
using ::crypto::tink::test::IsOk;
using ::crypto::tink::test::StatusIs;
using ::crypto::tink::util::StatusOr;
using ::testing::HasSubstr;

// Checks whether Decrypt(Encrypt(message)) == message with the given dem.
crypto::tink::util::Status EncryptThenDecrypt(
    const AeadOrDaead& dem, absl::string_view message,
    absl::string_view associated_data) {
  StatusOr<std::string> encryption_or = dem.Encrypt(message, associated_data);
  if (!encryption_or.status().ok()) return encryption_or.status();
  StatusOr<std::string> decryption_or =
      dem.Decrypt(encryption_or.ValueOrDie(), associated_data);
  if (!decryption_or.status().ok()) return decryption_or.status();
  if (decryption_or.ValueOrDie() != message) {
    return crypto::tink::util::Status(absl::StatusCode::kInternal,
                                      "Message/Decryption mismatch");
  }
  return util::OkStatus();
}

TEST(EciesAeadHkdfDemHelperTest, InvalidKey) {
  google::crypto::tink::KeyTemplate dem_key_template;
  dem_key_template.set_type_url("some.type.url/that.is.not.supported");
  auto result = EciesAeadHkdfDemHelper::New(dem_key_template);
  EXPECT_THAT(EciesAeadHkdfDemHelper::New(dem_key_template).status(),
              StatusIs(absl::StatusCode::kInvalidArgument,
                       HasSubstr("Unsupported DEM")));
}

TEST(EciesAeadHkdfDemHelperTest, DemHelperWithSomeAeadKeyType) {
  google::crypto::tink::AesGcmKeyFormat key_format;
  key_format.set_key_size(16);
  std::unique_ptr<AesGcmKeyManager> key_manager(new AesGcmKeyManager());
  std::string dem_key_type = key_manager->get_key_type();
  ASSERT_THAT(Registry::RegisterKeyTypeManager(std::move(key_manager), true),
              IsOk());

  google::crypto::tink::KeyTemplate dem_key_template;
  dem_key_template.set_type_url(dem_key_type);
  dem_key_template.set_value(key_format.SerializeAsString());

  auto dem_helper_or = EciesAeadHkdfDemHelper::New(dem_key_template);
  ASSERT_THAT(dem_helper_or.status(), IsOk());
  auto dem_helper = std::move(dem_helper_or.ValueOrDie());

  util::SecretData key128 = util::SecretDataFromStringView(
      test::HexDecodeOrDie("000102030405060708090a0b0c0d0e0f"));
  StatusOr<std::unique_ptr<AeadOrDaead>> aead_or_daead_result_or =
      dem_helper->GetAeadOrDaead(key128);
  ASSERT_THAT(aead_or_daead_result_or.status(), IsOk());

  auto aead_or_daead = std::move(aead_or_daead_result_or.ValueOrDie());
  EXPECT_THAT(EncryptThenDecrypt(*aead_or_daead, "test_plaintext", "test_ad"),
              IsOk());
}

TEST(EciesAeadHkdfDemHelperTest, DemHelperWithSomeDeterministicAeadKeyType) {
  google::crypto::tink::AesSivKeyFormat key_format;
  key_format.set_key_size(64);
  std::unique_ptr<AesSivKeyManager> key_manager(new AesSivKeyManager());
  std::string dem_key_type = key_manager->get_key_type();
  ASSERT_THAT(Registry::RegisterKeyTypeManager(std::move(key_manager), true),
              IsOk());

  google::crypto::tink::KeyTemplate dem_key_template;
  dem_key_template.set_type_url(dem_key_type);
  dem_key_template.set_value(key_format.SerializeAsString());

  auto dem_helper_or = EciesAeadHkdfDemHelper::New(dem_key_template);
  ASSERT_THAT(dem_helper_or.status(), IsOk());
  auto dem_helper = std::move(dem_helper_or.ValueOrDie());

  util::SecretData key128 = util::SecretDataFromStringView(test::HexDecodeOrDie(
      "000102030405060708090a0b0c0d0e0f000102030405060708090a0b0c0d0e0f00010203"
      "0405060708090a0b0c0d0e0f000102030405060708090a0b0c0d0e0f"));
  StatusOr<std::unique_ptr<AeadOrDaead>> aead_or_daead_result_or =
      dem_helper->GetAeadOrDaead(key128);
  ASSERT_THAT(aead_or_daead_result_or.status(), IsOk());

  auto aead_or_daead = std::move(aead_or_daead_result_or.ValueOrDie());
  EXPECT_THAT(EncryptThenDecrypt(*aead_or_daead, "test_plaintext", "test_ad"),
              IsOk());
}

}  // namespace
}  // namespace tink
}  // namespace crypto
